flexget/plugins/input/torznab.py

```diff
@@ -0,0 +1,292 @@
+from builtins import *  # noqa pylint: disable=unused-import, redefined-builtin
+
+from future.moves.urllib.parse import urlencode
+from past.utils import old_div
+from bs4 import BeautifulSoup, element
+import logging
+
+from flexget import plugin
+from flexget.entry import Entry
+from flexget.event import event
+from flexget.plugin import PluginError
+from flexget.components.sites.utils import torrent_availability
+
+log = logging.getLogger('torznab')
+
+
+class Torznab(object):
+    """Torznab search plugin
+
+    Handles searching for tv shows and movies, with fallback to simple query strings if these are not available.
+    """
+
+    @property
+    def schema(self):
+        """The schema of the plugin"""
+        schema = {
+            'type': 'object',
+            'properties': {
+                'apikey': {'type': 'string'},
+                'categories': {'type': 'array', 'items': {'type': 'integer'}, 'default': []},
+                'searcher': {'type': 'string', 'enum': ['movie', 'tv', 'tvsearch', 'search'], 'default': 'search'},
+                'website': {'type': 'string', 'format': 'url'}
+            },
+            'required': ['website', 'apikey'],
+            'additionalProperties': False
+        }
+        return schema
+
+    def search(self, task, entry, config=None):
+        """Search interface"""
+        self._setup(task, config)
+        params = {}
+        if self.params['t'] == 'movie':
+            params = self._gather_movie_params(entry)
+        elif self.params['t'] == 'tvsearch':
+            params = self._gather_tvsearch_params(entry)
+        if 'q' not in params.keys():
+            params['q'] = entry['title']
+        return self.create_entries_from_query(self._build_url(**params), task)
+
+    def _build_url(self, **kwargs):
+        """Builds the url with query parameters from the arguments"""
+        params = self.params.copy()
+        params.update(kwargs)
+        log.debug('Configured parameters: {}'.format(params))
+        url = '{}/api?'.format(self.base_url)
+        url = '{}{}'.format(url, urlencode(params))
+        return url
+
+    def _setup(self, task, config):
+        """Set up parameters"""
+        self.base_url = config['website'].rstrip('/')
+        self.supported_params = []
+        if config['searcher'] == 'tv':
+            config['searcher'] = 'tvsearch'
+
+        self.params = {
+            'apikey': config['apikey'],
+            'extended': 1,
+        }
+
+        log.debug('Config: {}'.format(config))
+        self._setup_caps(task, config['searcher'], config['categories'])
+
+    @plugin.internet(log)
+    def _setup_caps(self, task, searcher, categories):
+        """Gets the capabilities of the torznab indexer and matches it with the provided configuration"""
+
+        response = task.requests.get(self._build_url(t='caps'))
+        log.debug('Raw caps response {}'.format(response.content))
+        root = BeautifulSoup(response.content, 'lxml')
```

Comment:
What's wrong with html5lib?

Reply:
I've ran into issues with it not properly parsing everything, which as usual I can't reproduce now. I'll spend some time testing to see if I can get it to break again. If not I'll remove lxml.

Reply:
Figured it out, since htlm5lib is trying to parse the content as html not generic xml it does not parse self-closing tags as it is only accepting the ones defined in the html standard.

For example parsing the following with html5lib:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<caps>
  <server title="Jackett" />
  <searching>
    <search available="yes" supportedParams="q" />
    <tv-search available="yes" supportedParams="q,season,ep" />
    <movie-search available="no" supportedParams="q" />
    <music-search available="no" supportedParams="" />
    <audio-search available="no" supportedParams="" />
  </searching>
</caps>
```
Results in the `searching` element having children that make no sense:
```python
        for item in xml_root.find('searching').descendants:
            if type(item) == element.Tag:
                log.debug(item.prettify())
"""
<search available="yes" supportedparams="q">\n <tv-search available="yes" supportedparams="q,season,ep">\n  <movie-search available="no" supportedparams="q">\n   <music-search available="no" supportedparams="">\n    <audio-search available="no" supportedparams="">\n    </audio-search>\n   </music-search>\n  </movie-search>\n </tv-search>\n</search>
<tv-search available="yes" supportedparams="q,season,ep">\n <movie-search available="no" supportedparams="q">\n  <music-search available="no" supportedparams="">\n   <audio-search available="no" supportedparams="">\n   </audio-search>\n  </music-search>\n </movie-search>\n</tv-search>
<movie-search available="no" supportedparams="q">\n <music-search available="no" supportedparams="">\n  <audio-search available="no" supportedparams="">\n  </audio-search>\n </music-search>\n</movie-search>
<music-search available="no" supportedparams="">\n <audio-search available="no" supportedparams="">\n </audio-search>\n</music-search>
<audio-search available="no" supportedparams="">\n</audio-search>
"""
```
Compared to the expected output when parsed as xml
```python
"""
<search available="yes" supportedParams="q"/>\n
<tv-search available="yes" supportedParams="q,season,ep"/>\n
<movie-search available="no" supportedParams="q"/>\n
<music-search available="no" supportedParams=""/>\n
<audio-search available="no" supportedParams=""/>\n
"""
```

Manually passing a list of tags as self closing [is no longer supported](https://www.crummy.com/software/BeautifulSoup/bs4/doc/index.html?highlight=selfclosingtags#xml), that leaves using `lxml` or switching to `xml.etree.ElementTree` instead of bs4.

Let me know which you do you want me to proceed.

Reply:
What bothers me about lxml is that it depends on a C lib. ElementTree would be preferable.

@gazpachoking @paranoidi thoughts?

Reply:
I think we shouldn't add any deps to our requirements that aren't pure python. If lxml is the only way to get this plugin to work, I think we should make it an optional dependency and document it on the wiki. Installing lxml is a lot nicer these days too, since they provide wheels for a bunch of platforms on pypi.

Reply:
I'll just rewrite the xml parsing part to use ElementTree instead then.